iptables/src/utils.cpp

```diff
@@ -0,0 +1,309 @@
+static const std::string kIp6tablesNamesFile = "/proc/net/ip6_tables_names";
+
+static const std::string kHexMap = "0123456789ABCDEF";
+
+static const int kMaskHighBits = 4;
+static const int kMaskLowBits = 15;
+
+static TableList getTableNames(const std::string& filename) {
+  std::string content;
+
+  auto s = readFile(filename, content);
+
+  if (!s.ok()) {
+    TLOG << "Error reading: " << filename << ": " << s.toString();
+    return {};
+  }
+
+  auto results = SplitString(content, '\n');
+
+  for (auto& result : results) {
+    boost::algorithm::trim(result);
+  }
+
+  return results;
+}
+
+/* NOTE(ww): This function takes either iptables-save or ip6tables-save.
+ * It would be nice to use iptables-xml and ip6tables-xml, but ip6tables-xml
+ * doesn't exist yet and iptables-xml frequently hangs (during experimentation).
+ */
+Status genMatchMap(const std::string& ipt_cmd, MatchMap& match_map) {
+  ProcessOutput output;
+
+  if (!ExecuteProcess(output, ipt_cmd, {}) || output.exit_code != 0) {
+    return Status(1, "couldn't exec " + ipt_cmd);
+  }
+
+  if (output.std_output.empty()) {
+    return Status(1, "no output from command");
+  }
+
+  std::string table;
+  std::string chain;
+  for (const auto& line : SplitString(output.std_output, '\n')) {
+    std::string match;
+    std::string target;
+    std::string target_options;
+
+    // If the line is empty or a comment, skip it.
+    if (line.empty() || line.at(0) == '#') {
+      continue;
+    }
+
+    // If the line is a filter name, record it and prep our toplevel map with
+    // it.
+    if (line.at(0) == '*') {
+      table = line.substr(1);
+      match_map[table];
+      continue;
+    }
+
+    // If the line is a chain name, record it and prep our chain map with it.
+    if (line.at(0) == ':') {
+      auto stop = line.find(" ");
+      chain = line.substr(1, stop - 1);
+      match_map[table][chain];
+      continue;
+    }
+
+    // If the line is a rule, look for match entries, targets,
+    // and target options within it.
+    if (line.find("-A") == 0) {
+      auto start = line.find(" ");
+      auto stop = line.find(" ", start + 1);
+
+      chain = line.substr(start + 1, stop - start - 1);
+
+      // Matches begin with an -m.
+      start = line.find(" -m ");
+
+      // Targets begin with a -j or a -g.
+      stop = line.rfind(" -j ");
+      if (stop == std::string::npos) {
+        stop = line.rfind(" -g ");
+      }
+
+      // Match extraction.
+      if (start == std::string::npos) {
+        match = "";
+      } else {
+        if (stop != std::string::npos && stop < start) {
+          TLOG << "Oddity: -j or -g before -m: " << line;
+          match = "";
+        } else {
+          match = line.substr(start + 1, stop - start - 1);
+        }
+      }
+
+      // Target extraction.
+      if (stop == std::string::npos) {
+        target = "";
+      } else {
+        start = stop;
+        start = line.find(" ", start + 1);
```

Comment:
Yeah, something like `extractBetweenSpaces` perhaps. I didn't add a helper function before since the indices aren't always consistent (sometimes they're `start`, sometimes `start + 1`, sometimes `stop`, `stop - 1`, etc), but there's probably a way to simplify those too. Thanks!